File: backend/decision_engine/enrichment.py
"""
enrichment.py — Attach three contextual signals to each Mapbox route.

Signal 1 — congestion
  Average congestion level (0–100) of monitored locations within
  CORRIDOR_RADIUS_M metres of the route geometry.
  Source: PostGIS spatial query against traffic_data + locations tables.

  Note: Mapbox already bakes GPS-probe congestion into its ETA.
  This signal cross-validates with our own CPCB/IoT ground sensor network —
  different sources, different spatial coverage, different latency.
  Together they are more reliable than either alone.

Signal 2 — weather_penalty (0 → 1)
  Open-Meteo WMO code → penalty mapping. Computed at the route midpoint.

Signal 3 — stability (0 → 1)
  Geometric similarity to the current active route.
  0 = identical path, 1 = completely different road.
  Prevents switching for marginal improvements that add driver confusion.

All three run concurrently (asyncio.gather) — no serial waiting.
Each degrades gracefully: failure returns a neutral default, never raises.
"""
from __future__ import annotations

import asyncio
import logging
import math
from typing import Optional

import httpx
from sqlalchemy.orm import Session
from sqlalchemy import text

from .config import (
    weather_code_to_penalty,
    CORRIDOR_RADIUS_M,
    STABILITY_SAMPLE_N,
)
from .schemas import EnrichedRoute, MapboxRoute, RouteGeometry

log = logging.getLogger(__name__)


# ── Signal 1: PostGIS congestion ──────────────────────────────────────────────

async def get_congestion_for_route(
    geometry: RouteGeometry,
    db: Session,
) -> tuple[float, int]:
    """
    Returns (avg_congestion, readings_count).
    Falls back to (50.0, 0) if the query fails or returns nothing.
    """
    sql = text("""
        SELECT
            COALESCE(AVG(t.congestion_level), 50.0)::float  AS avg_congestion,
            COUNT(t.id)::int                                  AS readings_count
        FROM traffic_data t
        JOIN locations l ON l.id = t.location_id
        WHERE ST_DWithin(
            l.geom::geography,
            ST_GeomFromGeoJSON(:geom_json)::geography,
            :radius_m
        )
        AND t.timestamp > NOW() - INTERVAL '15 minutes'
    """)

    import json
    geom_json = json.dumps({"type": geometry.type, "coordinates": geometry.coordinates})

    try:
        result = db.execute(sql, {"geom_json": geom_json, "radius_m": CORRIDOR_RADIUS_M})
        row = result.mappings().first()
        if row:
            return float(row["avg_congestion"]), int(row["readings_count"])
        return 50.0, 0
    except Exception as exc:
        log.warning("[enrichment] PostGIS query failed: %s", exc)
        return 50.0, 0


# ── Signal 2: Weather penalty ─────────────────────────────────────────────────

async def get_weather_penalty(coordinates: list[list[float]]) -> float:
    """
    Fetches Open-Meteo current conditions at the route midpoint.
    Returns a 0→1 penalty. Returns 0.0 (assume clear) on any failure.
    """
    mid = coordinates[len(coordinates) // 2]
    lng, lat = mid[0], mid[1]   # GeoJSON is [lng, lat]

    url = (
        f"https://api.open-meteo.com/v1/forecast"
        f"?latitude={lat}&longitude={lng}"
        f"&current=weather_code"
        f"&timezone=auto"
    )

    try:
        async with httpx.AsyncClient(timeout=4.0) as client:
            resp = await client.get(url)
            resp.raise_for_status()
            data = resp.json()
            code = data.get("current", {}).get("weather_code", 0)
            return weather_code_to_penalty(int(code))
    except Exception as exc:
        log.warning("[enrichment] Weather fetch failed: %s", exc)
        return 0.0


# ── Signal 3: Stability score ─────────────────────────────────────────────────

def compute_stability(
    new_coords:     list[list[float]],
    current_coords: Optional[list[list[float]]],
) -> float:
    """
    Hausdorff-inspired mean minimum distance between sampled route points.
    Returns 0 when there is no current route (first trip = no switching cost).

    Algorithm:
      1. Sample N evenly-spaced points along each route
      2. For each new-route point, find the nearest current-route point
      3. Average those minimum distances
      4. Normalise: 0 km → 0.0, ≥ 5 km mean deviation → 1.0
    """
    if not current_coords or len(current_coords) == 0:
        return 0.0

    new_sampled     = _sample_evenly(new_coords,     STABILITY_SAMPLE_N)
    current_sampled = _sample_evenly(current_coords, STABILITY_SAMPLE_N)

    total_min_dist = 0.0
    for np_ in new_sampled:
        min_dist = min(_haversine_km(np_, cp) for cp in current_sampled)
        total_min_dist += min_dist

    mean_dev_km = total_min_dist / len(new_sampled)
    return min(1.0, mean_dev_km / 5.0)


def _sample_evenly(coords: list[list[float]], n: int) -> list[list[float]]:
    if len(coords) <= n:
        return coords
    step = (len(coords) - 1) / (n - 1)
    return [coords[round(i * step)] for i in range(n)]


def _haversine_km(p1: list[float], p2: list[float]) -> float:
    R = 6_371.0
    lng1, lat1 = p1[0], p1[1]
    lng2, lat2 = p2[0], p2[1]
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    a = (math.sin(d_lat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2))
         * math.sin(d_lng / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


# ── Main enrichment function ──────────────────────────────────────────────────

async def enrich_route(
    route:         MapboxRoute,
    current_route: Optional[object],  # CurrentRouteState | None
    db:            Session,
) -> EnrichedRoute:
    """
    Enriches a single Mapbox route with congestion, weather, and stability.
    PostGIS + weather calls run concurrently.
    """
    current_coords = (
        current_route.geometry.coordinates
        if current_route else None
    )

    # Fire PostGIS + weather concurrently — no reason to serialise
    (avg_congestion, readings_count), weather_penalty = await asyncio.gather(
        get_congestion_for_route(route.geometry, db),
        get_weather_penalty(route.geometry.coordinates),
    )

    stability = compute_stability(route.geometry.coordinates, current_coords)

    return EnrichedRoute(
        route_id         = route.route_id,
        duration         = route.duration,
        distance         = route.distance,
        geometry         = route.geometry,
        congestion       = avg_congestion,
        congestion_basis = readings_count,
        weather_penalty  = weather_penalty,
        stability        = stability,
    )


async def enrich_all_routes(
    routes:        list[MapboxRoute],
    current_route: Optional[object],
    db:            Session,
) -> list[EnrichedRoute]:
    """Enriches all routes concurrently."""
    return await asyncio.gather(
        *[enrich_route(r, current_route, db) for r in routes]
    )

File: lib/decisionEngine/enrichment.js
/**
 * enrichment.js — Attach real-world signals to each Mapbox route.
 *
 * Three signals are added to every route:
 *
 *   1. congestion  — Average congestion level (0–100) along the route corridor.
 *                    Source: PostGIS spatial query against our traffic_data table.
 *                    Note: Mapbox already bakes congestion into its ETA. This signal
 *                    cross-validates with our own ground sensor network (CPCB/IoT),
 *                    which has different coverage and update frequency.
 *
 *   2. weatherPenalty — Normalised 0→1 penalty based on current weather conditions.
 *                       Source: Open-Meteo free API (no key required).
 *
 *   3. stability   — How geometrically similar this route is to the active one.
 *                    0 = identical path, 1 = completely different road.
 *                    Prevents switching to a marginally faster route that adds
 *                    unnecessary cognitive load for the driver.
 *
 * The DB parameter accepts either:
 *   - A node-postgres Pool/Client instance (direct PostGIS access)
 *   - An object with a .query() method that proxies to the FastAPI backend
 *
 * All functions are pure async — no side effects, no global state.
 */

'use strict';

const { weatherCodeToPenalty, CORRIDOR_RADIUS_M, STABILITY_SAMPLE_N } = require('./config');

// ── Congestion from PostGIS ───────────────────────────────────────────────────

/**
 * Returns the average congestion level for locations within CORRIDOR_RADIUS_M
 * metres of the route geometry, based on readings from the last 15 minutes.
 *
 * Returns 50 (neutral) if no readings exist — never throws.
 *
 * @param {Object} routeGeometry - GeoJSON LineString from Mapbox route
 * @param {Object} db            - pg Pool/Client or compatible adapter
 * @returns {{ avgCongestion: number, readingsCount: number }}
 */
async function getCongestionForRoute(routeGeometry, db) {
  const sql = `
    SELECT
      COALESCE(AVG(t.congestion_level), 50)::float  AS avg_congestion,
      COUNT(t.id)::int                               AS readings_count
    FROM traffic_data t
    JOIN locations l ON l.id = t.location_id
    WHERE ST_DWithin(
      l.geom::geography,
      ST_GeomFromGeoJSON($1)::geography,
      $2
    )
    AND t.timestamp > NOW() - INTERVAL '15 minutes'
  `;

  try {
    const result = await db.query(sql, [
      JSON.stringify(routeGeometry),
      CORRIDOR_RADIUS_M,
    ]);
    const row = result.rows[0];
    return {
      avgCongestion:  parseFloat(row.avg_congestion) || 50,
      readingsCount:  parseInt(row.readings_count, 10) || 0,
    };
  } catch (err) {
    // Non-fatal — degrade gracefully, log for observability
    console.warn('[enrichment] PostGIS congestion query failed:', err.message);
    return { avgCongestion: 50, readingsCount: 0 };
  }
}

// ── Weather penalty from Open-Meteo ──────────────────────────────────────────

/**
 * Fetches current weather at the route midpoint and maps the WMO weather code
 * to a 0→1 penalty value.
 *
 * Uses the route's midpoint coordinate — a rough but fast approximation.
 * For very long routes consider sampling 3 points and averaging.
 *
 * @param {number[][]} coordinates - [[lng, lat], ...] from route geometry
 * @returns {number} penalty between 0 (clear) and 1 (storm)
 */
async function getWeatherPenalty(coordinates) {
  const mid  = coordinates[Math.floor(coordinates.length / 2)];
  const [lng, lat] = mid;   // GeoJSON is [lng, lat]

  const url = (
    `https://api.open-meteo.com/v1/forecast` +
    `?latitude=${lat}&longitude=${lng}` +
    `&current=weather_code,precipitation` +
    `&timezone=auto`
  );

  try {
    const res  = await fetch(url, { signal: AbortSignal.timeout(4_000) });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const data = await res.json();
    const code = data.current?.weather_code ?? 0;
    return weatherCodeToPenalty(code);
  } catch (err) {
    console.warn('[enrichment] Weather fetch failed:', err.message);
    return 0;   // Assume clear — never block routing on weather API failure
  }
}

// ── Stability score (geometric route similarity) ─────────────────────────────

/**
 * Computes how different the new route is from the currently active route.
 *
 * Algorithm:
 *   1. Sample STABILITY_SAMPLE_N evenly-spaced points along each route
 *   2. For each point on the new route, find the nearest point on the current route
 *   3. Average those minimum distances (Hausdorff-inspired, not exact)
 *   4. Normalise: 0 km → 0 (identical), ≥5 km mean deviation → 1 (completely different)
 *
 * Returns 0 when there is no current route (first trip, no penalty for change).
 *
 * @param {number[][]} newCoords     - [[lng,lat], ...] of candidate route
 * @param {number[][]|null} currentCoords - [[lng,lat], ...] of active route
 * @returns {number} 0–1 stability score
 */
function computeStabilityScore(newCoords, currentCoords) {
  if (!currentCoords || currentCoords.length === 0) return 0;

  const newSampled     = sampleEvenly(newCoords,     STABILITY_SAMPLE_N);
  const currentSampled = sampleEvenly(currentCoords, STABILITY_SAMPLE_N);

  let totalMinDist = 0;
  for (const np of newSampled) {
    let minDist = Infinity;
    for (const cp of currentSampled) {
      const d = haversineKm(np, cp);
      if (d < minDist) minDist = d;
    }
    totalMinDist += minDist;
  }

  const meanDevKm = totalMinDist / newSampled.length;

  // 0 km = same road, 5 km average deviation = completely different corridor
  return Math.min(1, meanDevKm / 5);
}

// ── Geometry helpers ──────────────────────────────────────────────────────────

/**
 * Samples N evenly-spaced points from an array of coordinates.
 * Handles edge cases: fewer points than N, or N=1.
 */
function sampleEvenly(coords, n) {
  if (coords.length <= n) return coords;
  const step = (coords.length - 1) / (n - 1);
  return Array.from({ length: n }, (_, i) => coords[Math.round(i * step)]);
}

/**
 * Haversine distance in kilometres between two [lng, lat] points.
 * Accurate enough for city-scale (~0.1% error over 50 km).
 */
function haversineKm([lng1, lat1], [lng2, lat2]) {
  const R    = 6371;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a    = (
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2
  );
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function toRad(deg) { return deg * (Math.PI / 180); }

// ── Main enrichment function ──────────────────────────────────────────────────

/**
 * Attaches congestion, weatherPenalty, and stability to a single Mapbox route.
 *
 * @param {Object} route         - Single Mapbox Directions route object
 * @param {Object|null} currentRoute - The route currently in use (or null)
 * @param {Object} db            - PostGIS adapter
 * @returns {Object} Enriched route with .congestion .weatherPenalty .stability
 */
async function enrichRoute(route, currentRoute, db) {
  const geometry  = route.geometry;          // GeoJSON LineString
  const coords    = geometry.coordinates;    // [[lng,lat], ...]

  const currentCoords = currentRoute?.geometry?.coordinates ?? null;

  // Run PostGIS + weather in parallel — no reason to serialise
  const [{ avgCongestion, readingsCount }, weatherPenalty] = await Promise.all([
    getCongestionForRoute(geometry, db),
    getWeatherPenalty(coords),
  ]);

  const stability = computeStabilityScore(coords, currentCoords);

  return {
    ...route,
    congestion:      avgCongestion,
    congestionBasis: readingsCount,   // How many readings backed this (0 = inferred)
    weatherPenalty,
    stability,
  };
}

module.exports = {
  enrichRoute,
  getCongestionForRoute,
  getWeatherPenalty,
  computeStabilityScore,
  haversineKm,
};
